src/Report.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE CPP #-}
+module Report (
+  Report
+, ReportState (..)
+, Summary(..)
+, report
+, report_
+) where
+
+#if __GLASGOW_HASKELL__ < 710
+import           Data.Monoid hiding ((<>))
+#endif
+
+import           Control.Monad hiding (forM_)
+import           Text.Printf (printf)
+import           System.IO (hPutStrLn, hPutStr, stderr)
+
+import           Control.Monad.Trans.State
+import           Control.Monad.IO.Class
+
+-- | Summary of a test run.
+data Summary = Summary {
+  sExamples :: Int
+, sTried    :: Int
+, sErrors   :: Int
+, sFailures :: Int
+} deriving Eq
+
+-- | Format a summary.
+instance Show Summary where
+  show (Summary examples tried errors failures) =
+    printf "Examples: %d  Tried: %d  Errors: %d  Failures: %d" examples tried errors failures
+
+
+-- | Sum up summaries.
+instance Monoid Summary where
+  mempty = Summary 0 0 0 0
+#if MIN_VERSION_base(4,11,0)
+instance Semigroup Summary where
+  (<>)
+#else
+  mappend
+#endif
+    (Summary x1 x2 x3 x4) (Summary y1 y2 y3 y4) = Summary (x1 + y1) (x2 + y2) (x3 + y3) (x4 + y4)
+
+-- | A monad for generating test reports.
+type Report = StateT ReportState IO
+
+data ReportState = ReportState {
+  reportStateCount        :: Int     -- ^ characters on the current line
+, reportStateInteractive  :: Bool    -- ^ should intermediate results be printed?
+, reportStateSummary      :: Summary -- ^ test summary
+}
+
+-- | Add output to the report.
+report :: String -> Report ()
+report msg = do
+  overwrite msg
+
+  -- add a newline, this makes the output permanent
+  liftIO $ hPutStrLn stderr ""
+  modify (\st -> st {reportStateCount = 0})
+
+-- | Add intermediate output to the report.
+--
+-- This will be overwritten by subsequent calls to `report`/`report_`.
+-- Intermediate out may not contain any newlines.
+report_ :: String -> Report ()
+report_ msg = do
+  f <- gets reportStateInteractive
+  when f $ do
+    overwrite msg
+    modify (\st -> st {reportStateCount = length msg})
+
+-- | Add output to the report, overwrite any intermediate out.
+overwrite :: String -> Report ()
+overwrite msg = do
+  n <- gets reportStateCount
+  let str | 0 < n     = "\r" ++ msg ++ replicate (n - length msg) ' '
+          | otherwise = msg
+  liftIO (hPutStr stderr str)
```

Comment:
Assuming `runProperty` can stay in `IO` then can this code be moved back where it was before (again, to minimize the diff and speed up reviewing)?

Reply:
Apparently it could not. And even if it could, it may be better for it to stay separate, so that reporting functionality was available for any other module.